### PR TITLE
[SYCL] Move splitting module code into a new file

### PR DESCRIPTION
**Pull request by mlychkov**
sycl-post-link tool's code should be grouped into several source files
to separate support of different functionalities, provided by the tool.
It should improve code readability and sustainability.
Current part of sycl-post-link tool code reorganization:
- Move splitting mechanism into a separate file.
- Make several ModuleSplitter classes to reduce comparisons during
  splitting.
- Modify types that handle module splitting to simplify migration to
  multithreading split in the future.

Signed-off-by: Mikhail Lychkov <mikhail.lychkov@intel.com>

### CODE REVIEW

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+  EntryPointGroupVec Groups = groupEntryPointsByKernelType(
+      *M, SplitEsimd, EmitOnlyKernelsAsEntryPoints);
+  bool DoSplit = (Groups.size() > 1);
+
+  std::unique_ptr<ModuleSplitterBase> Splitter;
+  if (!DoSplit)
+    Splitter.reset(new ModuleCopier(std::move(M), std::move(Groups)));
+  else
+    Splitter.reset(new ModuleSplitter(std::move(M), std::move(Groups)));
+  return Splitter;
```
Comment (psamolysov-intel) [inline llvm/tools/sycl-post-link/ModuleSplitter.cpp:406]:
```suggestion
  if (DoSplit)
    return std::make_unique<ModuleSplitter>(std::move(M), std::move(Groups));
  else
    return std::make_unique<ModuleCopier>(std::move(M), std::move(Groups));
```

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+    checkImageScopedDeviceGlobals(*M, Groups);
+  bool DoSplit = (Mode != SPLIT_NONE &&
+                  (Groups.size() > 1 || !Groups.cbegin()->Functions.empty()));
+
+  std::unique_ptr<ModuleSplitterBase> Splitter;
+  if (!DoSplit)
+    Splitter.reset(new ModuleCopier(std::move(M), std::move(Groups)));
+  else
+    Splitter.reset(new ModuleSplitter(std::move(M), std::move(Groups)));
+  return Splitter;
```
Comment (psamolysov-intel) [inline llvm/tools/sycl-post-link/ModuleSplitter.cpp:428]:
```suggestion
  if (DoSplit)
    return std::make_unique<ModuleSplitter>(std::move(M), std::move(Groups));
  else
    return std::make_unique<ModuleCopier>(std::move(M), std::move(Groups));
```

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+                  bool DeviceGlobals) {
+  EntryPointsGroupScope Scope =
+      selectDeviceCodeGroupScope(*M, Mode, IROutputOnly);
+  EntryPointGroupVec Groups =
+      groupEntryPointsByScope(*M, Scope, EmitOnlyKernelsAsEntryPoints);
+  assert(!Groups.empty() && "At least one group is expected");
+  if (DeviceGlobals)
+    checkImageScopedDeviceGlobals(*M, Groups);
+  bool DoSplit = (Mode != SPLIT_NONE &&
+                  (Groups.size() > 1 || !Groups.cbegin()->Functions.empty()));
```
Comment (psamolysov-intel) [inline llvm/tools/sycl-post-link/ModuleSplitter.cpp:421]: We are checking for at least a single function in the *first* group but why not in *every* one?

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+
+void collectFunctionsToExtract(SetVector<const GlobalValue *> &GVs,
+                               const EntryPointGroup &ModuleEntryPoints) {
+  for (const auto *F : ModuleEntryPoints.Functions)
+    GVs.insert(F);
+
+  decltype(GVs.size()) I = 0;
+  while (I != GVs.size()) {
+    const auto *F = cast<Function>(GVs[I++]);
+    for (const auto &I : instructions(F))
```
Comment (psamolysov-intel) [inline llvm/tools/sycl-post-link/ModuleSplitter.cpp:299]: This `I` shadows the first `I` what is an index. Maybe we should rename the first `I` in `Var` or `Idx`, or something like this?

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+      break;
+
+    case Scope_PerModule: {
+      if (!F.hasFnAttribute(ATTR_SYCL_MODULE_ID))
+        // TODO It may make sense to group all entry points w/o the attribute
+        // into a separate module rather than issuing an error. Should probably
+        // be controlled by an option.
+        error("no '" + Twine(ATTR_SYCL_MODULE_ID) +
+              "' attribute for entry point '" + F.getName() +
+              "', per-module split not possible");
```
Comment (psamolysov-intel) [inline llvm/tools/sycl-post-link/ModuleSplitter.cpp:202]:
```suggestion
              "', per-module split is not possible");
```

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 6]
```diff
@@ -616,9 +324,10 @@ void saveModuleIR(Module &M, StringRef OutFilename) {
   PrintModule.run(M);
 }
 
-void saveModuleProperties(Module &M, const EntryPointGroup &ModuleEntryPoints,
+void saveModuleProperties(module_split::ModuleDesc &ResM,
```
Comment (psamolysov-intel) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:327]: Do we change `ModuleDesc` here? If not (I can find where), should the reference be `const`?

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 7]
```diff
@@ -616,9 +324,10 @@ void saveModuleIR(Module &M, StringRef OutFilename) {
...
                           const GlobalBinImageProps &ImgPSInfo,
                           const std::string &PropSetFile) {
+  Module &M = ResM.getModule();
```
Comment (psamolysov-intel) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:330]:
```suggestion
  const Module &M = ResM.getModule();
```

**Review by psamolysov-intel — COMMENTED** [review]
Good job, thank you @mlychkov Some my comments are below. I'm not sure we should use nested namespaces in the `sycl-post-link` tool, it is not such big as the compiler driver, for example, but this is just an opinion.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 6 above)
Comment (mlychkov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:327]: We don't change it, but `MAM.getResult<SYCLKernelParamOptInfoAnalysis>(Module &M)` is called here which uses standard `ModuleAnalysisManager` interface.

File: llvm/tools/sycl-post-link/Utils.h [hunk 8]
```diff
@@ -0,0 +1,28 @@
+//===----------- Utils.h - utilities for sycl-post-link tool --------------===//
```
[the remark this answers is not in the document]
Comment (mlychkov) [reply llvm/tools/sycl-post-link/Utils.h]: Thank you. I just had older version of sycl-post-link code and didn't see that new file.
Moved `Utils.h` content into `Support.h` file.

File: llvm/tools/sycl-post-link/ModuleSplitter.h [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+protected:
+  const EntryPointGroup &nextGroup() {
+    assert(!EOL());
+    return *(GroupsIt++);
+  }
+
+  Module &getInputModule() { return *InputModule; }
+  std::unique_ptr<Module> releaseInputModule() {
+    return std::move(InputModule);
```
[the remark this answers is not in the document]
Comment (mlychkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.h:75]: Added assert check for methods accessing `InputModule` field.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 10]
```diff
@@ -108,7 +96,7 @@ cl::opt<bool> SplitEsimd{"split-esimd",
 
 // TODO Design note: sycl-post-link should probably separate different kinds of
 // its functionality on logical and source level:
-//  - LLVM IR module splitting
+//  - LLVM IR module splitting (done)
```
[the remark this answers is not in the document]
Comment (mlychkov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Removed the line.

**Review by psamolysov-intel — DISMISSED** [review]
LGTM

**Comment by mlychkov** [comment]
> I'm not sure we should use nested namespaces in the `sycl-post-link` tool, it is not such big as the compiler driver, for example, but this is just an opinion.

I have no strong opinion here regarding using nested namespaces or not. @AlexeySachkov what do you think?

**Comment by mlychkov** [comment]
> @mlychkov Why have you removed the check for the correct using of device global variables?

Sorry, I removed it by accident. Added that check back.

**Comment by mlychkov** [comment]
> I'm not sure the namespace prefix is required before a class, the class is already a "namespace" for its methods. So I see no such prefixes in `AliasAnalysis.cpp`, for example: `AAResults::AAResults(AAResults &&Arg)`, not `llvm::AAResults::AAResults(AAResults &&Arg)`. As I understand, the directive `using namespace llvm` takes all the reachable classes from the `llvm` namespace into the scope and lets us write the class name without the namespace prefix even in the `ClassName::method()` constructs (as well as we write class names for parameter and variable declarations, etc.)
> 
> As I get, the namespace prefix is required for free function definitions only.

Yes, thank you for prompt argumentation. I am agree with it. Removed llvm namespace prefix from member functions definitions.

File: llvm/tools/sycl-post-link/ModuleSplitter.h [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+  virtual ~ModuleSplitterBase() = default;
+
+  // Gets next subsequence of entry points in an input module and provides split
+  // submodule containing these entry points and their dependencies.
+  virtual ModuleDesc nextSplit() = 0;
+
+  size_t totalSplits() const { return Groups.size(); }
+
+  // Check that there are no more submodules to split.
+  bool EOL() const { return GroupsIt == Groups.cend(); }
```
Comment (kbobrovs) [inline llvm/tools/sycl-post-link/ModuleSplitter.h:96]: Nit: name acronym for "end of line/life", neither seem to fit the purpose. `hasMoreSplits`?

**Review by kbobrovs — DISMISSED** [review]
@mlychkov, just wanted to note that my invoke_simd patches depend on this refactoring, as one of the patches makes changes to splitting and post-link. None of my review comments are supposed to hold this PR, and can be addressed later.

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp [hunk 12]
```diff
@@ -0,0 +1,428 @@
...
+
+  case SPLIT_AUTO: {
+    if (IROutputOnly) {
+      // We allow enabling auto split mode even in presence of -ir-output-only
+      // flag, but in this case we are limited by it so we can't do any split
+      // at all.
+      return Scope_Global;
+    }
+
+    if (hasIndirectFunctionCalls(M))
```
[the remark this answers is not in the document]
Comment (mlychkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp]: There is already a comment inside that function.

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp [hunk 13]
```diff
@@ -0,0 +1,428 @@
...
+  // Kernels are always considered to be entry points
+  if (CallingConv::SPIR_KERNEL == F.getCallingConv())
+    return true;
+
+  if (!EmitOnlyKernelsAsEntryPoints) {
+    // If not disabled, SYCL_EXTERNAL functions with sycl-module-id attribute
+    // are also considered as entry points (except __spirv_* and __sycl_*
+    // functions)
+    return F.hasFnAttribute(ATTR_SYCL_MODULE_ID) &&
+           !isSpirvSyclBuiltin(F.getName());
```
[the remark this answers is not in the document]
Comment (mlychkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp:140]: I think it should be applied in an additional patch and have LIT tests for that cases.

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp [hunk 14]
```diff
@@ -0,0 +1,428 @@
...
+      }
+    };
+
+    SmallSetVector<const User *, 32> Workqueue;
+    for (auto *U : GV.users())
+      Workqueue.insert(U);
+
+    while (!Workqueue.empty()) {
+      const User *U = Workqueue.back();
+      Workqueue.pop_back();
```
[the remark this answers is not in the document]
Comment (mlychkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp]: Thank you

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp [hunk 15]
```diff
@@ -0,0 +1,428 @@
...
+        if (EntryPointModules.count(F))
+          CheckEntryPointModule(F);
+      }
+      for (auto *UU : U->users())
+        Workqueue.insert(UU);
+    }
+  }
+}
+
+void collectFunctionsToExtract(SetVector<const GlobalValue *> &GVs,
```
[the remark this answers is not in the document]
Comment (mlychkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp:293]: It is just for interface unification. CloneModule uses vector of `const GlobalValue *` values.

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp [hunk 16]
```diff
@@ -0,0 +1,428 @@
...
+    GVs.insert(F);
+
+  decltype(GVs.size()) Idx = 0;
+  while (Idx != GVs.size()) {
+    const auto *F = cast<Function>(GVs[Idx++]);
+    for (const auto &I : instructions(F))
+      if (const auto *CB = dyn_cast<CallBase>(&I))
+        if (const Function *CF = CB->getCalledFunction())
+          if (!CF->isDeclaration())
+            GVs.insert(CF);
```
[the remark this answers is not in the document]
Comment (mlychkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp:307]: Added a comment that this is not expected.

File: llvm/tools/sycl-post-link/ModuleSplitter.h (hunk 11 above)
Comment (mlychkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.h]: Thank you for suggestion. Applied.

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp [hunk 17]
```diff
@@ -0,0 +1,427 @@
...
+  using ModuleSplitterBase::ModuleSplitterBase; // to inherit base constructors
+
+  ModuleDesc nextSplit() override {
+    return extractCallGraph(getInputModule(), nextGroup());
+  }
+};
+
+} // namespace
+
+bool llvm::module_split::ModuleDesc::isEsimd() {
```
Comment (psamolysov-intel) [inline llvm/tools/sycl-post-link/ModuleSplitter.cpp:392]: Just a nit: it looks like the `llvm` namespace prefix can be removed here too in the same way as you did it for the `module_split::getSlitterByKernelType` function.

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (mlychkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp:140]: Made an issue for that improvement: https://github.com/intel/llvm/issues/5878

**Review by kbobrovs — APPROVED** [review]

**Comment by mlychkov** [comment]
Failures in OCL GEN9 and ESIMD Emu test suites are known XPASSes.
Failure in precommit test suite is infrastructure issue. Re-run has passed: http://icl-jenkins.sc.intel.com:8080/blue/organizations/jenkins/SYCL_CI%2Fintel%2FWin%2FTest_Suite/detail/Test_Suite/9916/pipeline/

**Review by psamolysov-intel — APPROVED** [review]
LGTM

**Comment by mlychkov** [comment]
Failed tests in ESIMD emulator:
  SYCL :: accessor_gather_scatter.cpp
  SYCL :: accessor_load_store.cpp
  SYCL :: api/esimd_pack_unpack_mask.cpp
  SYCL :: api/simd_copy_to_from.cpp
had failed in another PRs as well. However, they no more failed for the last 2 days.
I restarted tests to check if they are passed here as well.

**Pull request merged**